src/internal/operators/catchError.ts

```diff
@@ -111,44 +111,22 @@ export function catchError<T, O extends ObservableInput<any>>(
   return (source: Observable<T>) =>
     lift(source, function (this: Subscriber<T>, source: Observable<T>) {
       const subscriber = this;
-      const subscription = new Subscription();
-      let innerSub: Subscription | null = null;
-      let syncUnsub = false;
-      let handledResult: Observable<ObservedValueOf<O>>;
 
-      const handleError = (err: any) => {
-        try {
-          handledResult = from(selector(err, catchError(selector)(source)));
-        } catch (err) {
-          subscriber.error(err);
-          return;
-        }
-      };
-
-      innerSub = source.subscribe(
+      let subscription: Subscription | null = source.subscribe(
         new CatchErrorSubscriber(subscriber, (err) => {
-          handleError(err);
-          if (handledResult) {
-            if (innerSub) {
-              innerSub.unsubscribe();
-              innerSub = null;
-              subscription.add(handledResult.subscribe(subscriber));
-            } else {
-              syncUnsub = true;
-            }
+          try {
+            const nextSource = from(selector(err, catchError(selector)(source)));
+            subscription = nextSource.subscribe(subscriber);
```

Comment:
I actually hate that test with the `concat(defer, defer, defer)`. It should probably be just a plain observable with a loop, similar to "firehose" tests @cartant recently added.

Reply:
> I actually hate that test with the concat(defer, defer, defer). It should probably be just a plain observable with a loop, similar to "firehose" tests @cartant recently added.

Totally agree. In fact I'm pretty sure that test is redundant because unless I'm mistaken the "firehose" test is already covering that case, right?